commands/create/nodepool/command.go

```diff
@@ -0,0 +1,357 @@
+	}
+
+	if args.ClusterID == "" {
+		return microerror.Mask(errors.ClusterIDMissingError)
+	}
+
+	// AZ flags plausibility
+	if len(args.AvailabilityZonesList) > 0 && args.AvailabilityZonesNum > 0 {
+		return microerror.Maskf(errors.ConflictingFlagsError, "the flags --availability-zones and --num-availability-zones cannot be combined.")
+	}
+
+	// Scaling flags plausibility
+	if args.ScalingMin > 0 && args.ScalingMax > 0 {
+		if args.ScalingMin > args.ScalingMax {
+			return microerror.Mask(errors.WorkersMinMaxInvalidError)
+		}
+	}
+
+	return nil
+}
+
+func printValidation(cmd *cobra.Command, positionalArgs []string) {
+	var err error
+	args, err := collectArguments(positionalArgs)
+	if err == nil {
+		err = verifyPreconditions(args)
+	}
+
+	if err == nil {
+		return
+	}
+
+	errors.HandleCommonErrors(err)
+	client.HandleErrors(err)
+
+	headline := ""
+	subtext := ""
+
+	switch {
+	case errors.IsConflictingFlagsError(err):
+		headline = "Conflicting flags used"
+		subtext = "The flags --availability-zones and --num-availability-zones must not be used together."
+	default:
+		headline = err.Error()
+	}
+
+	// print output
+	fmt.Println(color.RedString(headline))
+	if subtext != "" {
+		fmt.Println(subtext)
+	}
+	os.Exit(1)
+}
+
+// createNodePool is the business function sending our creation request to the API
+// and returning either a proper result or an error.
+func createNodePool(args Arguments) (result, error) {
+	r := result{}
+
+	requestBody := &models.V5AddNodePoolRequest{
+		Name: args.Name,
+	}
+	if args.InstanceType != "" {
+		requestBody.NodeSpec = &models.V5AddNodePoolRequestNodeSpec{
+			Aws: &models.V5AddNodePoolRequestNodeSpecAws{
+				InstanceType: args.InstanceType,
+			},
+		}
+	}
+	if args.AvailabilityZonesList != nil && len(args.AvailabilityZonesList) > 0 {
+		requestBody.AvailabilityZones = &models.V5AddNodePoolRequestAvailabilityZones{
+			Zones: args.AvailabilityZonesList,
+		}
+	} else if args.AvailabilityZonesNum != 0 {
+		requestBody.AvailabilityZones = &models.V5AddNodePoolRequestAvailabilityZones{
+			Number: int64(args.AvailabilityZonesNum),
+		}
+	}
+	if args.ScalingMin != 0 || args.ScalingMax != 0 {
+		requestBody.Scaling = &models.V5AddNodePoolRequestScaling{
+			Min: args.ScalingMin,
+			Max: args.ScalingMax,
+		}
+	}
+
+	if args.Verbose {
+		fmt.Println(color.WhiteString("Submitting node pool creation reqwuest with the following details:"))
+		fmt.Println(color.WhiteString("%+v\n", requestBody))
+	}
+
+	clientWrapper, err := client.NewWithConfig(args.APIEndpoint, args.AuthToken)
+	if err != nil {
+		return r, microerror.Mask(err)
+	}
+
+	auxParams := clientWrapper.DefaultAuxiliaryParams()
+	auxParams.ActivityName = activityName
+
+	response, err := clientWrapper.CreateNodePool(args.ClusterID, requestBody, auxParams)
+
+	if err != nil {
+		// create specific error types for cases we care about
+		if clientErr, ok := err.(*clienterror.APIError); ok {
+			if clientErr.HTTPStatusCode == http.StatusForbidden {
```

Comment:
Please see how we deal with errors in `fmt`. We can have custom error matchers for this.

Reply:
Added and applied error matchers to the `clienterror` package.

Reply:
I see. Though, the error matcher interface is supposed to take `error`. Our `fmt` guideline is pretty clear about this. The type assertion should happen within the matcher, not outside of it. And then the error handling is nice and flat like everywhere else.

Reply:
See the example for [`IsAlreadyTerminated`](https://github.com/giantswarm/aws-operator/blob/58efc26ee0f8d34b00527aa583a0c27cbd7996b3/service/controller/clusterapi/v29/resource/tccp/error.go#L23-L48) in the `aws-operator`.

Reply:
Thanks for the example! I moved the type assertion into the matchers.